### ISSUE TITLE
Integration test that focus on AMM informing our decision to toggle this on by default

### ISSUE THREAD

**Issue by fjetter** [coiled/coiled-runtime#140]

**Comment by crusaderky** [comment]
We should scale this up to production size:

https://github.com/dask/distributed/blob/f7f650154fea29978906c65dd0225415da56ed11/distributed/tests/test_active_memory_manager.py#L1079-L1085
https://github.com/dask/distributed/blob/f7f650154fea29978906c65dd0225415da56ed11/distributed/tests/test_active_memory_manager.py#L1111-L1130

**Comment by crusaderky** [comment]
Between this issue and #135, there are four use cases to be implemented:

- just run tensordot_stress on production_size data
- tensordot_stress with AMM ReduceReplicas turned on
- tensordot_stress with AMM turned off and retiring 90% of the cluster while the computation is going
- tensordot_stress with AMM ReduceReplicas turned on and retiring 90% of the cluster while the computation is going

**Comment by crusaderky** [comment]
# DOD / AC

This story is done when the integration test portrays the behaviour of distributed on coiled as described above.
If it demonstrates a flaw, remediation of the flaw is out of scope.

**Comment by fjetter** [comment]
I agree with the above, this ticket should exclusively focus on ReduceReplicas. Questions I would like to be answered

- Is enabling this by default impacting common workflows? Are any regressions/improvements detected in existing coiled-runtime benchmarks
- Are there any relevant missing benchmarks we would like to be added to the benchmark suite
- There should be an upstream ticket presenting the results of the benchmarking efforts to propose enabling this by default

Optional
- The ReduceReplicas is currently "aggressively" set to one replica. I'm wondering if setting this to a less aggressive value would even accelerate us further.

**Comment by crusaderky** [comment]
> * The ReduceReplicas is currently "aggressively" set to one replica. I'm wondering if setting this to a less aggressive value would even accelerate us further.

It's set to one replica or the number of waiters, whatever is higher:
https://github.com/dask/distributed/blob/a80187a23bd317136d5dbbb02068bc8f90055e4e/distributed/active_memory_manager.py#L494-L502

**Comment by crusaderky** [comment]
# Executive summary

- The latest A/B replication machinery with statistical analysis (#324) suffers from major levels of background noise, which can be mitigated by throwing money at the problem through a very high number of repeat runs (>15).
Failing that, a lower number of repeat runs (7, in this case study) will result in very noisy graphics, which could easily mislead an untrained eye and should not be shown as-is to the general public.
- Enabling AMM ReduceReplicas across the whole spectrum of tests caused:
  - no noticeable changes in terms of runtime
  - no noticeable regressions in terms of memory usage
  - improvements up to 30% in terms of memory usage, both average and peak, in some specific use cases.

# Test setup
### Baseline
- python=3.9
- coiled-runtime=0.1.0
- dask=2022.9.0
- distributed=2022.9.0
- benchmarks only (no runtime, no stability)
- repeat every test 7 times
- test null hypothesis

### AMM
Everything as baseline, with one difference:
```yaml
distributed:
  scheduler:
    active-memory-manager:
      start: true
```
The above enables AMM ReduceReplicas to run every 2 seconds.

# Observations on noise (unrelated to AMM)
Given the current volatility in the outcomes of the test suite, 7 runs are not enough to produce obviously clear results. The null hypothesis contains several major flares that are not supposed to be there (see pictures below). Cross-comparison with the barcharts and timeseries reports shows how these flares are caused by one-off instances where either runtime took twice as much wall clock time, twice as much memory, or both.

Examples:
- #315
- #339
- #340

Given a flare that happens 1~3 times over 7 repeats, it's intuitive to understand that a runtime where the flare happened 3 times will be reported in the A/B plots as a major regression when compared to a runtime where it happened only once, and that only a much higher number of repeats would prevent false positives.

Everything displayed in these plots is just noise:
![image](https://user-images.githubusercontent.com/6213168/190629091-75e76d4d-7c15-4e05-907a-997114c2da5f.png)
![image](https://user-images.githubusercontent.com/6213168/190629271-5fe35117-553e-4405-96e5-ba4f3b12df9b.png)
![image](https://user-images.githubusercontent.com/6213168/190629334-5dbc77ff-9be1-42ab-964e-578c6332eab1.png)


# AMM-specific observations
Net of noise, the runs with AMM enabled highlight some changes, which can be generally distinguished from noise by the length of their tail - a spread-out flare with a long tail is typical of a high-variance measure, whereas measures that are in solid color (very high p-value) followed by a sudden drop denote a low-variance change.

![image](https://user-images.githubusercontent.com/6213168/190632671-b239a406-a3c8-4c59-9d64-72c54b6e192d.png)

## Enhancements
| Test name | Average memory (p>90%) | Peak memory (p>90%) |
| --- | --- | --- |
| h2o/test_hd2_benchmarks.py::test_q3[5 GB (parquet)] | -10% to -15% | -15% to -20% |
| h2o/test_hd2_benchmarks.py::test_q7[5 GB (parquet)] | -10% to -15% | -10% to -15% |
| h2o/test_hd2_benchmarks.py::test_q8[5 GB (parquet)] | -25% to -30% | -25% to -30% |
| test_array.py::test_dot_product | -10% to -15% | -10% to -15% |
| test_dataframe.py::test_shuffle | -5% to -10% | -5% to -10% |

![image](https://user-images.githubusercontent.com/6213168/190635454-3d44b83e-561c-4939-94be-ecdd72e76b42.png)
![image](https://user-images.githubusercontent.com/6213168/190635602-1daae7ce-cc9d-4359-8503-1d68c927445a.png)
![image](https://user-images.githubusercontent.com/6213168/190635656-d568bc10-630b-4d14-823a-f8c60c9fa451.png)
![image](https://user-images.githubusercontent.com/6213168/190634342-f7cd8bfe-830c-494a-8039-c89b4cb44599.png)
![image](https://user-images.githubusercontent.com/6213168/190635298-1d169195-4207-4dcc-a1b8-879db2b6164c.png)


## Noise
![image](https://user-images.githubusercontent.com/6213168/190636171-57c545b6-5e76-479a-8c61-1306a1d36688.png)

For test_download_throughput[pandas], refer to #339.

**Comment by crusaderky** [comment]
CC @fjetter @hendrikmakait @hayesgb @mrocklin @jrbourbeau

**Comment by fjetter** [comment]
Thank you, @crusaderky ! I believe this is a pretty convincing result. Would you do us the honor and open a PR to distributed to make this official?

We can also see some significant improvements in Wall Clock for something like test_vorticity (this one is working under a lot of memory pressure and is spilling a lot)

![image](https://user-images.githubusercontent.com/8629629/190638259-a47df9bc-c3e4-4956-adbb-4577f6f9a355.png)

**Comment by ntabris** [comment]
My understanding is that in many of the benchmarks place stress on dask such that we wouldn't expect consistent performance. Is that correct? Are there cases were we would currently expect to see consistent performance but are seeing higher than expected variance / occasional flares? (Or is that not yet determined?)

**Comment by fjetter** [comment]
> My understanding is that in many of the benchmarks place stress on dask such that we wouldn't expect consistent performance. Is that correct? 

Yes. For some of these tests we're addressing this in https://github.com/coiled/coiled-runtime/pull/338

> Are there cases were we would currently expect to see consistent performance but are seeing higher than expected variance / occasional flares? (Or is that not yet determined?)

I think https://github.com/coiled/coiled-runtime/issues/339 is an example of "not yet understood" (or rather not confirmed, yet)
